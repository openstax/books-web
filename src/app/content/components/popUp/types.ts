import { LinkedArchiveTree, LinkedArchiveTreeNode, LinkedArchiveTreeSection } from '../../types';

export type LocationFilters = Map<
  string,
  { section: LinkedArchiveTreeNode, children?: LinkedArchiveTreeSection[] }
>;

export type LocationFiltersWithChildren = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;

export interface FiltersChange<T> {
  remove: T[];
  new: T[];
}
