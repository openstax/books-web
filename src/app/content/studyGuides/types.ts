import { HighlightsSummary } from '@openstax/highlighter/dist/api';

export interface State {
  isEnabled: boolean;
  summary: HighlightsSummary | null;
}
