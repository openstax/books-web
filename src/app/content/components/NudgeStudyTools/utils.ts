import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides, studyGuidesEnabled } from '../../studyGuides/selectors';
import {
  arrowDesktopHeight,
  arrowLeftMargin,
  arrowMobileHeight,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  contentMarginTop,
  cookieNudge,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  spotlightPadding,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperRight: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const getPositions = (target: HTMLElement, isMobile: boolean, windowWidth: number): Positions => {
  const { top, left, right, height, width } = getBoundingRectOfNudgeTarget(target);
  const padding = remsToPx(spotlightPadding);
  const spotlightTopOffset = top - padding;
  const spotlightLeftOffset = left - padding;
  const spotlightHeight = height + (padding * 2);
  const spotlightWidth = width + (padding * 2);
  const arrowTopOffset = spotlightTopOffset + spotlightHeight + remsToPx(arrowTopMargin);
  const arrowLeft = spotlightLeftOffset + remsToPx(arrowLeftMargin);
  const contentWrapperTopOffset = arrowTopOffset
    + remsToPx(isMobile ? arrowMobileHeight : arrowDesktopHeight)
    + remsToPx(contentMarginTop);
  const contentWrapperRight = windowWidth - right - padding;
  const closeButtonLeft = left + spotlightWidth;
  const closeButtonTopOffset = contentWrapperTopOffset - remsToPx(closeButtonDistanceFromContent);

  return {
    arrowLeft,
    arrowTopOffset,
    closeButtonLeft,
    closeButtonTopOffset,
    contentWrapperRight,
    contentWrapperTopOffset,
    spotlightHeight,
    spotlightLeftOffset,
    spotlightTopOffset,
    spotlightWidth,
  };
};

// Target may have display set to `contents` so we are calculatings rect for children of passed target.
const getBoundingRectOfNudgeTarget = (target: HTMLElement) => {
  // Starting values depends on if we checks for Math.max or Math.min
  let top = 9999;
  let bottom = 0;
  let left = 9999;
  let right = 0;

  for (const child of Array.from(target.children)) {
    const rect = child.getBoundingClientRect();
    top = Math.min(top, rect.top);
    bottom = Math.max(bottom, rect.bottom);
    left = Math.min(left, rect.left);
    right = Math.max(right, rect.right);
  }

  return {
    height: bottom - top,
    left,
    right,
    top,
    width: right - left,
  };
};

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const studyGuides = useSelector(hasStudyGuides);
  const isEnabled = useSelector(studyGuidesEnabled);

  React.useEffect(() => {
    const document = assertDocument();
    const target = document.querySelector(`#${nudgeStudyToolsTargetId}`) as HTMLElement | null;
    if (isEnabled && target) {
      // Make sure that we calculate positions with body overflow set to hidden
      // because it causes scrollbar to hide which results with different positions.
      const prevOverflow = document.body.style.overflow;
      document.body.style.overflow = 'hidden';
      setPositions(getPositions(target, isMobile, windowWidth));

      // Resets to the value from before calculations. We want this style change to be handled
      // directly in the component.
      document.body.style.overflow = prevOverflow;
    }
    return () => setPositions(null);
  }, [studyGuides, isEnabled, windowWidth, isMobile]);

  return positions;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.counter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudge.date);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (Date.now() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.pageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  Cookies.set(cookieNudge.pageCounter, (counter + 1).toString());
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

// Set required cookies and reset opened page counter
export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudge.counter, (counter + 1).toString());
  Cookies.set(cookieNudge.date, now.toString());
  Cookies.remove(cookieNudge.pageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  return React.useEffect(() => {
    const counter = getPageCounterCookie();
    if (page && counter <= nudgeStudyToolsMinPageLimit) { incrementPageCounterCookie(); }
  }, [page]);
};
