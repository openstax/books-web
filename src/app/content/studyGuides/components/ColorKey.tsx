import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import colorKeyIcon from '../../../../assets/colorKeyIcon.png';
import { useAnalyticsEvent } from '../../../../helpers/analytics';
import { PlainButton } from '../../../components/Button';
import { textRegularStyle } from '../../../components/Typography';
import theme from '../../../theme';
import { toolbarIconStyles } from '../../components/Toolbar/iconStyles';
import { toolbarDefaultText } from '../../components/Toolbar/styled';
import ColorIndicator from '../../highlights/components/ColorIndicator';
import { mobilePaddingSides } from '../../styles/PopupConstants';
import { highlightStyles } from '../constants';

const noteConstants = {
  noteHeight: 16.8,
  noteWidth: 36,
};

// tslint:disable-next-line:variable-name
export const ColorKeyButtonWrapper = styled(PlainButton)`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: auto;
  padding-right: 2.4rem;
  position: relative;
  overflow: visible;
`;

// tslint:disable-next-line:variable-name
const ColorKeyDescription = styled.div`
  ::before {
    content: "";
    position: absolute;
    width: 1.6rem;
    height: 1.6rem;
    transform: rotate(45deg);
    top: -0.9rem;
    right: 5.5rem;
    z-index: 1;
    background: ${theme.color.white};
    border-top: 0.1rem solid ${theme.color.neutral.formBorder};
    border-left: 0.1rem solid ${theme.color.neutral.formBorder};

    ${theme.breakpoints.mobile(css`
      right: 2rem;
    `)}
  }

  min-height: ${noteConstants.noteHeight}rem;
  width: ${noteConstants.noteWidth}rem;
  box-sizing: border-box;
  border: 0.1rem solid ${theme.color.neutral.formBorder};
  box-shadow: 0 0.4rem 0.6rem 0 rgba(0,0,0,0.2);
  position: absolute;
  top: 4rem;
  right: 5rem;
  background: ${theme.color.white};
  overflow: visible;
  padding: 1.5rem 1.65rem;

  ${theme.breakpoints.mobile(css`
    width: calc(100vw - ${mobilePaddingSides * 2}rem);
    right: ${mobilePaddingSides / 2}rem;
  `)}
`;

// tslint:disable-next-line:variable-name
const ColorKeyIcon = styled.img`
  ${toolbarIconStyles}
`;

// tslint:disable-next-line:variable-name
const ColorKeyText = styled.span`
  ${toolbarDefaultText}
`;

// tslint:disable-next-line:variable-name
const KeyTermText = styled.span`
  ${textRegularStyle}
  font-size: 1.4rem;
  letter-spacing: -0.4px;
  margin-left: 1.6rem;
  width: 100%;
  text-align: left;
`;

// tslint:disable-next-line:variable-name
const KeyTermWrapper = styled.div`
  display: flex;
  align-items: flex-start;
  padding: 0.5rem 0;

  ${ColorIndicator} {
    margin-top: 0.5rem;
    width: 1.8rem;
  }
`;

// tslint:disable-next-line:variable-name
const ColorKey = () => {
  const [open, setOpen] = React.useState<boolean>(false);
  const trackOpenClose = useAnalyticsEvent('openCloseColorKey');

  const toggleColorKey = () => {
    setOpen((state) => !state);
    trackOpenClose(open);
  };

  return <FormattedMessage id='i18n:studyguides:popup:color-key'>
    {(msg: Element | string) =>
      <ColorKeyButtonWrapper onClick={toggleColorKey} aria-label={msg}>
        <ColorKeyIcon src={colorKeyIcon}/>
        <ColorKeyText>{msg}</ColorKeyText>
        {open && <ColorKeyDescription>
          {highlightStyles.map((color, index) =>
            <KeyTermWrapper key={index}>
              <FormattedMessage id='i18n:studyguides:popup:color-key:terms:aria-label' values={{color: color.label}}>
                {(arialabel: string) => <ColorIndicator
                  shape='circle'
                  style={color}
                  size='small'
                  aria-label={arialabel}
                  component={<label />}
                />}
              </FormattedMessage>
              <FormattedMessage id={'i18n:studyguides:popup:color-key:terms:' + color.label}>
                {(key: string) => <KeyTermText>{key}</KeyTermText>}
              </FormattedMessage>
            </KeyTermWrapper>
          )
        }
        </ColorKeyDescription>}
      </ColorKeyButtonWrapper>
    }
  </FormattedMessage>;
};

export default ColorKey;
