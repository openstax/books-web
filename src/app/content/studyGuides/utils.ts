import { HighlightColorEnum } from '@openstax/highlights-client';
import { OutputParams } from 'query-string';
import { colorFilterQueryParameterName } from '../constants';
import { colorfilterLabels } from './constants';

export const getColorFiltersFromQuery = (query: OutputParams) => {
  const queryColors = query[colorFilterQueryParameterName] as HighlightColorEnum | HighlightColorEnum[] | undefined;
  if (!queryColors) { return; }

  const colors = (Array.isArray(queryColors) ? queryColors : [queryColors])
    .filter((color) => colorfilterLabels.has(color));

  return colors.length ? colors : undefined;
};
