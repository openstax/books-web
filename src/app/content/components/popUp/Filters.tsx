import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import { AngleDown } from 'styled-icons/fa-solid/AngleDown';
import { PlainButton } from '../../../components/Button';
import Dropdown, { DropdownToggle } from '../../../components/Dropdown';
import { textStyle } from '../../../components/Typography/base';
import theme from '../../../theme';
import { filters } from '../../styles/PopupConstants';
import { disablePrint } from '../utils/disablePrint';
import FiltersList from './FiltersList';

// tslint:disable-next-line:variable-name
export const DownIcon = styled(AngleDown)`
  color: ${theme.color.primary.gray.base};
  width: ${filters.dropdownToggle.icon.width}rem;
  height: ${filters.dropdownToggle.icon.height}rem;
  margin-left: 0.8rem;
  padding-top: 0.2rem;
  ${(props: { rotate: boolean }) => {
    if (props.rotate) {
      return `
        transform: rotate(180deg);
        padding-top: 0;
        padding-bottom: 0.2rem;
      `;
    }
  }}
`;

interface ToggleProps {
  label: string;
  isOpen: boolean;
  ariaLabelId: string;
}
// tslint:disable-next-line:variable-name
const Toggle = styled(React.forwardRef<HTMLButtonElement, ToggleProps>(
  ({label, isOpen, ariaLabelId, ...props}, ref) => (
    <FormattedMessage id={ariaLabelId} values={{filter: label}}>
      {(msg: string) => <PlainButton ref={ref} {...props} aria-label={msg}>
        <div tabIndex={-1}>
          {label}
          <DownIcon rotate={isOpen} />
        </div>
      </PlainButton>}
    </FormattedMessage>
  )
))`
  position: relative;
  border-left: ${filters.border}rem solid transparent;
  border-right: ${filters.border}rem solid transparent;
  ${(props: ToggleProps) => props.isOpen
    ? css`
      z-index: 2;
      box-shadow: 0 0 0.6rem 0 rgba(0,0,0,0.2);
      clip-path: inset(0 -0.6rem 0px -0.6rem);
      background-color: ${theme.color.white};
      border-left: ${filters.border}rem solid ${theme.color.neutral.formBorder};
      border-right: ${filters.border}rem solid ${theme.color.neutral.formBorder};
    `
    : null
  }
  > div {
    padding: ${filters.dropdownToggle.topBottom.desktop}rem ${filters.dropdownToggle.sides.desktop}rem;
    ${theme.breakpoints.mobile(css`
      padding: ${filters.dropdownToggle.topBottom.mobile}rem ${filters.dropdownToggle.sides.mobile}rem;
    `)}
    outline: none;
    ${textStyle}
    font-size: 1.6rem;
    color: ${theme.color.primary.gray.base};
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: flex-end;
  }
`;

// tslint:disable-next-line:variable-name
export const FilterDropdown = ({label, ariaLabelId, children, closeWhenThisPropChange}:
  React.PropsWithChildren<{label: string, ariaLabelId: string, closeWhenThisPropChange?: any}>) =>
    <FormattedMessage id={label}>
      {(msg: Element | string) => <Dropdown
        toggle={<Toggle label={msg} ariaLabelId={ariaLabelId} />}
        transparentTab={false}
        closeWhenThisPropChange={closeWhenThisPropChange}
      >
        {children}
      </Dropdown>}
    </FormattedMessage>;

// tslint:disable-next-line: variable-name
export const FiltersTopBar = styled.div`
  display: flex;
  align-items: center;
  overflow: visible;
  width: 100%;
`;

interface Props {
  className?: string;
}

// tslint:disable-next-line:variable-name
const Filters = ({className, children}: React.PropsWithChildren<Props>) => {
  return <div className={className}>
    {children}
  </div>;
};

export default styled(Filters)`
  position: relative;
  z-index: 2;
  overflow: visible;
  display: flex;
  flex-flow: row wrap;
  align-items: center;
  background: ${theme.color.neutral.base};
  border-bottom: ${filters.border}rem solid ${theme.color.neutral.formBorder};
  ${css`
    ${DropdownToggle} {
      font-weight: bold;
    }

    ${Dropdown} {
      & > *:not(${DropdownToggle}) {
        top: calc(100% - ${filters.border}rem);
        box-shadow: 0 0 0.6rem 0 rgba(0, 0, 0, 0.2);
        max-height: calc(100vh - ${filters.valueToSubstractFromVH.desktop}rem);
        ${theme.breakpoints.mobile(css`
          max-height: calc(100vh - ${filters.valueToSubstractFromVH.mobile}rem);
        `)}
      }

      ${theme.breakpoints.mobileSmall(css`
        position: initial;

        & > *:not(${DropdownToggle}) {
          top: auto;
          margin-top: -${filters.border}rem;
        }
      `)}
    }
  `}

  @media print {
    padding-left: 0;
  }

  ${css`
    > *:not(${FiltersList}) {
      ${disablePrint}
    }
  `}
`;
