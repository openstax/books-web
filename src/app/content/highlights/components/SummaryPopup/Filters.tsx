import flow from 'lodash/fp/flow';
import React from 'react';
import { connect } from 'react-redux';
import { AppState, Dispatch } from '../../../../types';
import ChapterFilter from '../../../elements/popUp/ChapterFilter';
import Filters, { FilterDropdown } from '../../../elements/popUp/Filters';
import FiltersList from '../../../elements/popUp/FiltersList';
import PrintButton from '../../../elements/popUp/PrintButton';
import { printSummaryHighlights, setSummaryFilters } from '../../actions';
import * as select from '../../selectors';
import ColorFilter from './ColorFilter';

// tslint:disable-next-line:variable-name
export const ConnectedChapterFilter = connect(
  (state: AppState) => ({
    locationFilters: select.highlightLocationFilters(state),
    locationFiltersWithContent: select.highlightLocationFiltersWithContent(state),
    selectedLocationFilters: select.summaryLocationFilters(state),
  }),
  (dispatch: Dispatch) => ({
    setFilters: flow(setSummaryFilters, dispatch),
  })
)(ChapterFilter);

// tslint:disable-next-line:variable-name
export const ConnectedFilterList = connect(
  (state: AppState) => ({
    locationFilters: select.highlightLocationFilters(state),
    selectedColorFilters: select.summaryColorFilters(state),
    selectedLocationFilters: select.summaryLocationFilters(state),
  }),
  (dispatch: Dispatch) => ({
    setFilters: flow(setSummaryFilters, dispatch),
  })
)(FiltersList);

// tslint:disable-next-line:variable-name
export const ConnectedPrintButton = connect(
  (state: AppState) => ({
    disabled: select.summaryIsLoading(state),
    loading: select.summaryIsLoading(state),
    shouldFetchMore: select.hasMoreResults(state),
  }),
  (dispatch: Dispatch) => ({
    loadHighlightsAndPrint: flow(printSummaryHighlights, dispatch),
  }),
  (stateProps, dispatchProps, ownProps) => {
    const {shouldFetchMore, loadHighlightsAndPrint, ...props} = {
      ...stateProps,
      ...dispatchProps,
      ...ownProps,
    };

    return shouldFetchMore
      ? {...props, onClick: loadHighlightsAndPrint}
      : props
    ;
  }
)(PrintButton);

export default () =>
  <Filters>
    <FilterDropdown
      label='i18n:highlighting:filters:chapters'
      ariaLabelId='i18n:highlighting:filters:filter-by:aria-label'
    >
      <ConnectedChapterFilter />
    </FilterDropdown>
    <FilterDropdown
      label='i18n:highlighting:filters:colors'
      ariaLabelId='i18n:highlighting:filters:filter-by:aria-label'
    >
      <ColorFilter />
    </FilterDropdown>
    <ConnectedPrintButton />
    <ConnectedFilterList />
  </Filters>;
