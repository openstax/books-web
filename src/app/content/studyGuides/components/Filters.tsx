import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import flow from 'lodash/fp/flow';
import React from 'react';
import { connect, useSelector } from 'react-redux';
import styled from 'styled-components/macro';
import { loggedOut } from '../../../auth/selectors';
import { AppState, Dispatch } from '../../../types';
import ChapterFilter from '../../components/popUp/ChapterFilter';
import ColorFilter from '../../components/popUp/ColorFilter';
import Filters, { FilterDropdown, FiltersTopBar } from '../../components/popUp/Filters';
import FiltersList from '../../components/popUp/FiltersList';
import PrintButton from '../../components/popUp/PrintButton';
import { printStudyGuides, setSummaryFilters } from '../actions';
import { highlightStyles } from '../constants';
import * as selectors from '../selectors';
import UsingThisGuideBanner from './UsingThisGuide/UsingThisGuideBanner';
import UsingThisGuideButton from './UsingThisGuide/UsingThisGuideButton';

// tslint:disable-next-line:variable-name
const ConnectedChapterFilter = connect(
  (state: AppState) => ({
    locationFilters: selectors.studyGuidesLocationFilters(state),
    locationFiltersWithContent: selectors.studyGuidesLocationFiltersWithContent(state),
    selectedLocationFilters: selectors.summaryLocationFilters(state),
  }),
  (dispatch: Dispatch) => ({
    setFilters: flow(setSummaryFilters, dispatch),
  })
)(ChapterFilter);

// tslint:disable-next-line: variable-name
const StyledColorFilter = styled(ColorFilter)`
  min-width: 29rem;
`;

// tslint:disable-next-line: variable-name
const RightButtonsWrapper = styled.div`
  margin-left: auto;
  display: flex;
  align-items: center;
  overflow: visible;
`;

// tslint:disable-next-line: variable-name
const ConnectedColorFilter = connect(
  (state: AppState) => ({
    colorFiltersWithContent: selectors.highlightColorFiltersWithContent(state),
    selectedColorFilters: selectors.summaryColorFilters(state),
  }),
  (dispatch: Dispatch) => ({
    setSummaryFilters: flow(setSummaryFilters, dispatch),
  })
)(StyledColorFilter);

// tslint:disable-next-line:variable-name
const ConnectedFilterList = connect(
  (state: AppState) => ({
    locationFilters: selectors.studyGuidesLocationFilters(state),
    selectedColorFilters: selectors.summaryColorFilters(state),
    selectedLocationFilters: selectors.summaryLocationFilters(state),
  }),
  (dispatch: Dispatch) => ({
    setFilters: flow(setSummaryFilters, dispatch),
  })
)(FiltersList);

// tslint:disable-next-line:variable-name
const ConnectedPrintButton = connect(
  (state: AppState) => ({
    disabled: selectors.summaryIsLoading(state),
    loading: selectors.summaryIsLoading(state),
    shouldFetchMore: selectors.hasMoreResults(state),
  }),
  (dispatch: Dispatch) => ({
    loadHighlightsAndPrint: flow(printStudyGuides, dispatch),
  }),
  (stateProps, dispatchProps, ownProps) => {
    const {shouldFetchMore, loadHighlightsAndPrint, ...props} = {
      ...stateProps,
      ...dispatchProps,
      ...ownProps,
    };

    return shouldFetchMore
      ? {...props, onClick: loadHighlightsAndPrint}
      : props
    ;
  }
)(PrintButton);

export default () => {
  const userLoggedOut = useSelector(loggedOut);
  const [isUTGopen, setUTGopen] = React.useState(false);

  const toggleUsingThisGuide = () => {
    setUTGopen((state) => !state);
  };

  return <Filters>
    <FiltersTopBar>
      <FilterDropdown
        label='i18n:highlighting:filters:chapters'
        ariaLabelId='i18n:studyguides:popup:filters:filter-by:aria-label'
      >
        <ConnectedChapterFilter disabled={userLoggedOut}/>
      </FilterDropdown>
      <FilterDropdown
        label='i18n:highlighting:filters:colors'
        ariaLabelId='i18n:studyguides:popup:filters:filter-by:aria-label'
      >
        <ConnectedColorFilter
          disabled={userLoggedOut}
          styles={highlightStyles}
          labelKey={(label: HighlightColorEnum) => `i18n:studyguides:popup:filters:${label}`}
        />
      </FilterDropdown>
      <RightButtonsWrapper>
        <ConnectedPrintButton studyGuidesButton />
        <UsingThisGuideButton onClick={toggleUsingThisGuide} open={isUTGopen}/>
      </RightButtonsWrapper>
    </FiltersTopBar>
    { isUTGopen && <UsingThisGuideBanner onClick={toggleUsingThisGuide} isUTGopen={isUTGopen}/> }
    {!userLoggedOut && <ConnectedFilterList
      colorAriaLabelKey={() => 'i18n:studyguides:popup:filters:remove:color'}
      colorLabelKey={(label: HighlightColorEnum) => `i18n:studyguides:popup:filters:${label}`}
    />}
  </Filters>;
};
