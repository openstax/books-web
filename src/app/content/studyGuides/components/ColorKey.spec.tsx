import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import * as Services from '../../../context/Services';
import MessageProvider from '../../../MessageProvider';
import { Store } from '../../../types';
import { assertDocument } from '../../../utils';
import ColorKey, { ColorKeyButtonWrapper, ColorKeyDescription } from './ColorKey';
import Filters from './Filters';

describe('Study Guides button and PopUp', () => {
  let store: Store;
  let services: ReturnType<typeof createTestServices>;

  beforeEach(() => {
    services = createTestServices();
    store = createTestStore();
  });

  it('matches snapshot', () => {
    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <ColorKey />
        </MessageProvider>
      </Services.Provider>
    </Provider>);

    renderer.act(() => {
      const button = component.root.findByType(ColorKeyButtonWrapper);
      button.props.onClick();
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Opens and closes from click outside or button', () => {
    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <Filters />
        </MessageProvider>
      </Services.Provider>
    </Provider>, {createNodeMock: () => assertDocument().createElement('div')});

    const clickOutside = () => {
      const document = assertDocument();
      const elementOutside = document.createElement('div');
      document.body.appendChild(elementOutside);
      const event = document.createEvent('MouseEvent');
      event.initEvent('click', true, false);
      elementOutside.dispatchEvent(event);
      elementOutside.remove();
    };

    const button = component.root.findByType(ColorKeyButtonWrapper);

    /* open and close from clicking button*/
    renderer.act(() => button.props.onClick());
    expect(() => component.root.findByType(ColorKeyDescription)).not.toThrow();
    renderer.act(() => button.props.onClick());
    expect(() => component.root.findByType(ColorKeyDescription)).toThrow();

    /*close from clicking outside*/
    renderer.act(() => button.props.onClick());
    expect(() => component.root.findByType(ColorKeyDescription)).not.toThrow();
    renderer.act(() => clickOutside());
    expect(() => component.root.findByType(ColorKeyDescription)).toThrow();
  });
});
