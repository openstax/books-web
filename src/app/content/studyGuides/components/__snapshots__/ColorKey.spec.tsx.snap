// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Study Guides button and PopUp matches snapshot 1`] = `
<div
  className="ColorKey__ColorKeyWrapper-afl82f-6 dlLowB"
>
  <button
    aria-label="Color key"
    className="Button__PlainButton-ayg7nk-2 ColorKey__ColorKeyButtonWrapper-afl82f-0 fkFtKM"
    onClick={[Function]}
  >
    <img
      className="ColorKey__ColorKeyIcon-afl82f-2 kbHiQJ"
      src="colorKeyIcon.png"
    />
    <span
      className="ColorKey__ColorKeyText-afl82f-3 xtfbY"
    >
      Color key
    </span>
  </button>
  <div
    className="ColorKey__ColorKeyDescription-afl82f-1 dBMRUv"
  >
    <div
      className="ColorKey__KeyTermWrapper-afl82f-5 kwvQBa"
    >
      <label
        aria-label="yellow key term"
        className="ColorIndicator-sc-1w5zl7g-1 kBHzhz"
        shape="circle"
      >
        <svg
          aria-hidden="true"
          className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 512 512"
        >
          <path
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            fill="currentColor"
          />
        </svg>
      </label>
      <span
        className="ColorKey__KeyTermText-afl82f-4 dOnBcA"
      >
        Key terms and phrases
      </span>
    </div>
    <div
      className="ColorKey__KeyTermWrapper-afl82f-5 kwvQBa"
    >
      <label
        aria-label="green key term"
        className="ColorIndicator-sc-1w5zl7g-1 dSRHOr"
        shape="circle"
      >
        <svg
          aria-hidden="true"
          className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 512 512"
        >
          <path
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            fill="currentColor"
          />
        </svg>
      </label>
      <span
        className="ColorKey__KeyTermText-afl82f-4 dOnBcA"
      >
        Key connections, relationships between topics
      </span>
    </div>
    <div
      className="ColorKey__KeyTermWrapper-afl82f-5 kwvQBa"
    >
      <label
        aria-label="blue key term"
        className="ColorIndicator-sc-1w5zl7g-1 eibKCM"
        shape="circle"
      >
        <svg
          aria-hidden="true"
          className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 512 512"
        >
          <path
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            fill="currentColor"
          />
        </svg>
      </label>
      <span
        className="ColorKey__KeyTermText-afl82f-4 dOnBcA"
      >
        Key facts and summaries to know for tests/essays
      </span>
    </div>
    <div
      className="ColorKey__KeyTermWrapper-afl82f-5 kwvQBa"
    >
      <label
        aria-label="purple key term"
        className="ColorIndicator-sc-1w5zl7g-1 ekjDWZ"
        shape="circle"
      >
        <svg
          aria-hidden="true"
          className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 512 512"
        >
          <path
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            fill="currentColor"
          />
        </svg>
      </label>
      <span
        className="ColorKey__KeyTermText-afl82f-4 dOnBcA"
      >
        Enrichment
      </span>
    </div>
  </div>
</div>
`;
